File: app/src/main/java/com/haishinkit/rtmp/RTMPConnection.java
package com.haishinkit.rtmp;

import java.net.URI;
import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Map;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.ConcurrentHashMap;

import com.haishinkit.events.Event;
import com.haishinkit.events.EventDispatcher;
import com.haishinkit.events.IEventListener;
import com.haishinkit.lang.IRawValue;
import com.haishinkit.net.IResponder;
import com.haishinkit.rtmp.messages.RTMPCommandMessage;
import com.haishinkit.rtmp.messages.RTMPMessage;
import com.haishinkit.rtmp.messages.RTMPSetPeerBandwidthMessage;
import com.haishinkit.util.EventUtils;
import com.haishinkit.util.Log;

import org.apache.commons.lang3.StringUtils;

public class RTMPConnection extends EventDispatcher {
    public static final int DEFAULT_PORT = 1935;
    public static final String DEFAULT_FLASH_VER = "FMLE/3.0 (compatible; FMSc/1.0)";
    public static final RTMPObjectEncoding DEFAULT_OBJECT_ENCODING = RTMPObjectEncoding.AMF0;

    private static final int DEFAULT_CHUNK_SIZE_S = 1024 * 8;
    private static final int DEFAULT_CAPABILITIES = 239;

    public enum Codes implements IRawValue<String> {
        CALL_BAD_VERSION("NetConnection.Call.BadVersion", "error"),
        CALL_FAILED("NetConnection.Call.Failed", "error"),
        CALL_PROHIBITED("NetConnection.Call.Prohibited", "error"),
        CONNECT_APP_SHUTDOWN("NetConnection.Connect.AppShutdown", "status"),
        CONNECT_CLOSED("NetConnection.Connect.Closed", "status"),
        CONNECT_FAILED("NetConnection.Connect.Failed", "error"),
        CONNECT_IDLE_TIME_OUT("NetConnection.Connect.IdleTimeOut", "status"),
        CONNECT_INVALID_APP("NetConnection.Connect.InvalidApp", "error"),
        CONNECT_NETWORK_CHANGE("NetConnection.Connect.NetworkChange", "status"),
        CONNECT_REJECTED("NetConnection.Connect.Rejected", "status"),
        CONNECT_SUCCESS("NetConnection.Connect.Success", "status");

        private final String rawValue;
        private final String level;

        Codes(final String rawValue, final String level) {
            this.rawValue = rawValue;
            this.level = level;
        }

        public Map<String, Object> data(final String description) {
            Map<String, Object> data = new HashMap<String, Object>();
            data.put("code", rawValue);
            data.put("level", level);
            if (StringUtils.isNoneEmpty(description)) {
                data.put("description", description);
            }
            return data;
        }

        public String getLevel() {
            return level;
        }

        public String rawValue() {
            return rawValue;
        }
    }

    public enum SupportSound implements IRawValue<Short> {
        NONE((short) 0x001),
        ADPCM((short) 0x002),
        MP3((short) 0x004),
        INTEL((short) 0x008),
        UNUSED((short) 0x0010),
        NELLY8((short) 0x0020),
        NELLY((short) 0x0040),
        G711A((short) 0x0080),
        G711U((short) 0x0100),
        AAC((short) 0x0200),
        SPEEX((short) 0x0800),
        ALL((short) 0x0FFF);

        private final short rawValue;

        SupportSound(final short rawValue) {
            this.rawValue = rawValue;
        }

        public Short rawValue() {
            return rawValue;
        }
    }

    public enum SupportVideo implements IRawValue<Short> {
        UNUSED((short) 0x001),
        JPEG((short) 0x002),
        SORENSON((short) 0x004),
        HOMEBREW((short) 0x008),
        VP6((short) 0x0010),
        VP6_ALPHA((short) 0x0020),
        HOMEBREWV((short) 0x0040),
        H264((short) 0x0080),
        ALL((short) 0x00FF);

        private final short rawValue;

        SupportVideo(final short rawValue) {
            this.rawValue = rawValue;
        }

        public Short rawValue() {
            return this.rawValue;
        }
    }

    public enum VideoFunction implements IRawValue<Short> {
        CLIENT_SEEK((short) 1);

        private final short rawValue;

        VideoFunction(final short rawValue) {
            this.rawValue = rawValue;
        }

        public Short rawValue() {
            return this.rawValue;
        }
    }

    private final class EventListener implements IEventListener {

        private final RTMPConnection connection;

        private EventListener(final RTMPConnection connection) {
            this.connection = connection;
        }

        @Override
        public void handleEvent(final Event event) {
            Map<String, Object> data = EventUtils.toMap(event);
            switch (data.get("code").toString()) {
                case "NetConnection.Connect.Success":
                    connection.getSocket().setChunkSizeS(RTMPConnection.DEFAULT_CHUNK_SIZE_S);
                    connection.getSocket().doOutput(RTMPChunk.ONE,
                            new RTMPSetPeerBandwidthMessage()
                                    .setSize(RTMPConnection.DEFAULT_CHUNK_SIZE_S)
                                    .setChunkStreamID(RTMPChunk.CONTROL)
                    );
                    break;
                default:
                    break;
            }
        }
    }

    private URI uri = null;
    private int chunkSizeC = 0;
    private String swfUrl = null;
    private String pageUrl = null;
    private String flashVer = RTMPConnection.DEFAULT_FLASH_VER;
    private RTMPObjectEncoding objectEncoding = RTMPConnection.DEFAULT_OBJECT_ENCODING;
    private int transactionID = 0;
    private Object[] arguments = null;
    private Map<Short, ByteBuffer> payloads = new ConcurrentHashMap<Short, ByteBuffer>();
    private Map<Short, RTMPMessage> messages = new ConcurrentHashMap<Short, RTMPMessage>();
    private Map<Integer, RTMPStream> streams = new ConcurrentHashMap<Integer, RTMPStream>();
    private Map<Integer, IResponder> responders = new ConcurrentHashMap<Integer, IResponder>();
    private RTMPSocket socket = new RTMPSocket(this);

    public RTMPConnection() {
        super(null);
        addEventListener(Event.RTMP_STATUS, new EventListener(this));
    }

    public Map<Integer, RTMPStream> getStreams() {
        return streams;
    }

    public RTMPSocket getSocket() {
        return socket;
    }

    public RTMPObjectEncoding getObjectEncoding() {
        return objectEncoding;
    }

    public RTMPConnection setObjectEncoding(final RTMPObjectEncoding objectEncoding) {
        this.objectEncoding = objectEncoding;
        return this;
    }

    public URI getUri() {
        return uri;
    }

    public boolean isConnected() {
        return socket.isConnected();
    }

    public String getSwfUrl() {
        return swfUrl;
    }

    public RTMPConnection setSwfUrl(String swfUrl) {
        this.swfUrl = swfUrl;
        return this;
    }

    public String getPageUrl() {
        return pageUrl;
    }

    public RTMPConnection setPageUrl(String pageUrl) {
        this.pageUrl = pageUrl;
        return this;
    }

    public String getFlashVer() {
        return this.flashVer;
    }

    public RTMPConnection setFlashVer(String flashVer) {
        this.flashVer = flashVer;
        return this;
    }

    public Map<Integer, IResponder>  getResponders() {
        return responders;
    }

    public void call(final String commandName, final IResponder responder, final Object... arguments) {
        if (!isConnected()) {
            return;
        }
        List<Object> listArguments = new ArrayList<Object>(arguments.length);
        for (Object object : arguments) {
            listArguments.add(object);
        }
        RTMPCommandMessage message = new RTMPCommandMessage(objectEncoding);
        message.setChunkStreamID(RTMPChunk.COMMAND);
        message.setStreamID(0);
        message.setTransactionID(++transactionID);
        message.setCommandName(commandName);
        message.setArguments(listArguments);
        if (responder != null) {
            getResponders().put(transactionID, responder);
        }
        getSocket().doOutput(RTMPChunk.ZERO, message);
    }

    public void connect(final String command, final Object... arguments) {
        uri = URI.create(command);
        if (isConnected() || !uri.getScheme().equals("rtmp")) {
            return;
        }
        int port = uri.getPort();
        this.arguments = arguments;
        socket.connect(uri.getHost(), port == -1 ? RTMPConnection.DEFAULT_PORT : port);
    }

    public void close() {
        if (!isConnected()) {
            return;
        }
        socket.close();
    }

    Map<Short, RTMPMessage> getMessages() {
        return messages;
    }

    void listen(final ByteBuffer buffer) {

        int rollback = buffer.position();

        try {
            byte first = buffer.get();
            int chunkSizeC = socket.getChunkSizeC();
            RTMPChunk chunk = RTMPChunk.rawValue((byte) ((first & 0xff) >> 6));
            short streamID = chunk.getStreamID(buffer);

            ByteBuffer payload;
            RTMPMessage message;
            if (chunk == RTMPChunk.THREE) {
                payload = payloads.get(streamID);
                message = messages.get(streamID);
                int remaining = payload.remaining();
                if (chunkSizeC < remaining) {
                    remaining = chunkSizeC;
                }
                payload.put(buffer.array(), buffer.position(), remaining);
                buffer.position(buffer.position() + remaining);
                if (!payload.hasRemaining()) {
                    payload.flip();
                    message.decode(payload).execute(this);
                    Log.v(getClass().getName(), message.toString());
                    payloads.remove(payload);
                }
            } else {
                message = chunk.decode(streamID, this, buffer);
                if (message.getLength() <= chunkSizeC) {
                    message.decode(buffer).execute(this);
                    Log.v(getClass().getName(), message.toString());
                } else {
                    payload = ByteBuffer.allocate(message.getLength());
                    payload.put(buffer.array(), buffer.position(), chunkSizeC);
                    buffer.position(buffer.position() + chunkSizeC);
                    payloads.put(streamID, payload);
                }
                messages.put(streamID, message);
            }
        } catch (IndexOutOfBoundsException e) {
            buffer.position(rollback);
            throw e;
        }

        if (buffer.hasRemaining()) {
            listen(buffer);
        }
    }

    void createStream(final RTMPStream stream) {
        call("createStream", new IResponder() {
            @Override
            public void onResult(List<Object> arguments) {
                int id = new Double((double) arguments.get(0)).intValue();
                stream.setId(id);
                streams.put(id, stream);
                stream.setReadyState(RTMPStream.ReadyState.OPEN);
            }
            @Override
            public void onStatus(List<Object> arguments) {
                Log.w(getClass().getName() + "#onStatus", "");
            }
        });
    }

    RTMPMessage createConnectionMessage() {
        String[] paths = uri.getPath().split("/", 0);
        RTMPCommandMessage message = new RTMPCommandMessage(RTMPObjectEncoding.AMF0);
        Map<String, Object> commandObject = new HashMap<String, Object>();
        commandObject.put("app", paths[1]);
        commandObject.put("flashVer", getFlashVer());
        commandObject.put("swfUrl", getSwfUrl());
        commandObject.put("tcUrl", uri.toString());
        commandObject.put("fpad", false);
        commandObject.put("capabilities", RTMPConnection.DEFAULT_CAPABILITIES);
        commandObject.put("audioCodecs", SupportSound.AAC.rawValue());
        commandObject.put("videoCodecs", SupportVideo.H264.rawValue());
        commandObject.put("videoFunction", VideoFunction.CLIENT_SEEK.rawValue());
        commandObject.put("pageUrl", getPageUrl());
        commandObject.put("objectEncoding", objectEncoding.rawValue());
        message.setChunkStreamID(RTMPChunk.COMMAND);
        message.setStreamID(0);
        message.setCommandName("connect");
        message.setTransactionID(++transactionID);
        message.setCommandObject(commandObject);
        if (arguments != null) {
            message.setArguments(Arrays.asList(arguments));
        }
        return message;
    }
}

File: haishinkit/src/main/java/com/haishinkit/rtmp/RTMPConnection.kt
package com.haishinkit.rtmp

import android.util.Log
import com.haishinkit.events.Event
import com.haishinkit.events.EventDispatcher
import com.haishinkit.events.IEventListener
import com.haishinkit.net.Responder
import com.haishinkit.rtmp.messages.RTMPCommandMessage
import com.haishinkit.rtmp.messages.RTMPMessage
import com.haishinkit.rtmp.messages.RTMPMessageFactory
import com.haishinkit.util.EventUtils
import org.apache.commons.lang3.StringUtils
import java.net.URI
import java.nio.ByteBuffer
import java.util.Timer
import java.util.TimerTask
import java.util.concurrent.ConcurrentHashMap
import kotlin.concurrent.schedule

open class RTMPConnection : EventDispatcher(null) {

    enum class Code(val rawValue: String, val level: String) {
        CALL_BAD_VERSION("NetConnection.Call.BadVersion", "error"),
        CALL_FAILED("NetConnection.Call.Failed", "error"),
        CALL_PROHIBITED("NetConnection.Call.Prohibited", "error"),
        CONNECT_APP_SHUTDOWN("NetConnection.Connect.AppShutdown", "status"),
        CONNECT_CLOSED("NetConnection.Connect.Closed", "status"),
        CONNECT_FAILED("NetConnection.Connect.Failed", "error"),
        CONNECT_IDLE_TIME_OUT("NetConnection.Connect.IdleTimeOut", "status"),
        CONNECT_INVALID_APP("NetConnection.Connect.InvalidApp", "error"),
        CONNECT_NETWORK_CHANGE("NetConnection.Connect.NetworkChange", "status"),
        CONNECT_REJECTED("NetConnection.Connect.Rejected", "status"),
        CONNECT_SUCCESS("NetConnection.Connect.Success", "status");

        fun data(description: String): Map<String, Any> {
            val data = HashMap<String, Any>()
            data.put("code", rawValue)
            data.put("level", level)
            if (StringUtils.isNoneEmpty(description)) {
                data.put("description", description)
            }
            return data
        }
    }

    enum class SupportSound(val rawValue: Short) {
        NONE(0x001),
        ADPCM(0x002),
        MP3(0x004),
        INTEL(0x008),
        UNUSED(0x0010),
        NELLY8(0x0020),
        NELLY(0x0040),
        G711A(0x0080),
        G711U(0x0100),
        AAC(0x0200),
        SPEEX(0x0800),
        ALL(0x0FFF);
    }

    enum class SupportVideo(val rawValue: Short) {
        UNUSED(0x001),
        JPEG(0x002),
        SORENSON(0x004),
        HOMEBREW(0x008),
        VP6(0x0010),
        VP6_ALPHA(0x0020),
        HOMEBREWV(0x0040),
        H264(0x0080),
        ALL(0x00FF);
    }

    enum class VideoFunction(val rawValue: Short) {
        CLIENT_SEEK(1);
    }

    private inner class EventListener(private val connection: RTMPConnection) : IEventListener {
        override fun handleEvent(event: Event) {
            val data = EventUtils.toMap(event)
            when (data["code"].toString()) {
                RTMPConnection.Code.CONNECT_SUCCESS.rawValue -> {
                    timerTask = Timer().schedule(0, 1000) {
                        for (stream in streams) stream.value.on()
                    }
                    var message = messageFactory.createRTMPSetChunkSizeMessage()
                    message.size = RTMPConnection.DEFAULT_CHUNK_SIZE_S
                    message.chunkStreamID = RTMPChunk.CONTROL
                    connection.socket.chunkSizeS = RTMPConnection.DEFAULT_CHUNK_SIZE_S
                    connection.doOutput(RTMPChunk.ZERO, message)
                }
            }
        }
    }

    var uri: URI? = null
        private set
    var swfUrl: String? = null
    var pageUrl: String? = null
    var flashVer = RTMPConnection.DEFAULT_FLASH_VER
    var objectEncoding = RTMPConnection.DEFAULT_OBJECT_ENCODING
    internal val messages = ConcurrentHashMap<Short, RTMPMessage>()
    internal val streams = ConcurrentHashMap<Int, RTMPStream>()
    internal val responders = ConcurrentHashMap<Int, Responder>()
    internal val socket = RTMPSocket(this)
    internal var transactionID = 0
    internal val messageFactory = RTMPMessageFactory(4)
    private var timerTask: TimerTask? = null
        set(value) {
            timerTask?.cancel()
            field = value
        }
    private var arguments: MutableList<Any?> = mutableListOf()
    private val payloads = ConcurrentHashMap<Short, ByteBuffer>()

    init {
        addEventListener(Event.RTMP_STATUS, EventListener(this))
    }

    val isConnected: Boolean
        get() = socket.isConnected

    fun call(commandName: String, responder: Responder?, vararg arguments: Any) {
        if (!isConnected) {
            return
        }
        val listArguments = ArrayList<Any>(arguments.size)
        for (`object` in arguments) {
            listArguments.add(`object`)
        }
        val message = RTMPCommandMessage(objectEncoding)
        message.chunkStreamID = RTMPChunk.COMMAND
        message.streamID = 0
        message.transactionID = ++transactionID
        message.commandName = commandName
        message.arguments = listArguments
        if (responder != null) {
            responders[transactionID] = responder
        }
        doOutput(RTMPChunk.ZERO, message)
    }

    fun connect(command: String, vararg arguments: Any?) {
        uri = URI.create(command)
        if (isConnected || uri!!.scheme != "rtmp") {
            return
        }
        val port = uri!!.port
        this.arguments.clear()
        arguments.forEach { value ->
            this.arguments.add(value)
        }
        socket.connect(uri!!.host, if (port == -1) RTMPConnection.DEFAULT_PORT else port)
    }

    fun close() {
        if (!isConnected) {
            return
        }
        timerTask = null
        socket.close(false)
    }

    fun dispose() {
        timerTask = null
        streams.forEach {
            it.value.dispose()
        }
        streams.clear()
    }

    internal fun doOutput(chunk: RTMPChunk, message: RTMPMessage) {
        for (buffer in chunk.encode(socket, message)) {
            socket.doOutput(buffer)
        }
        messageFactory.release(message)
    }

    internal fun listen(buffer: ByteBuffer) {
        val rollback = buffer.position()
        try {
            val first = buffer.get()
            val chunkSizeC = socket.chunkSizeC
            var chunk = RTMPChunk.values().first { v -> v.rawValue.toInt() == ((first.toInt() and 0xff) shr 6) }
            val streamID = chunk.getStreamID(buffer)
            val payload: ByteBuffer
            val message: RTMPMessage
            if (chunk == RTMPChunk.THREE) {
                payload = payloads[streamID]!!
                message = messages[streamID]!!
                var remaining = payload.remaining()
                if (chunkSizeC < remaining) {
                    remaining = chunkSizeC
                }
                payload.put(buffer.array(), buffer.position(), remaining)
                buffer.position(buffer.position() + remaining)
                if (!payload.hasRemaining()) {
                    payload.flip()
                    message.decode(payload).execute(this)
                    messageFactory.release(message)
                    Log.v(javaClass.name + "#listen", message.toString())
                    payloads.remove(streamID)
                }
            } else {
                message = chunk.decode(streamID, this, buffer)
                if (message.length <= chunkSizeC) {
                    message.decode(buffer).execute(this)
                    messageFactory.release(message)
                    Log.v(javaClass.name + "#listen", message.toString())
                } else {
                    payload = ByteBuffer.allocate(message.length)
                    payload.put(buffer.array(), buffer.position(), chunkSizeC)
                    buffer.position(buffer.position() + chunkSizeC)
                    payloads[streamID] = payload
                }
                messages[streamID] = message
            }
        } catch (e: IndexOutOfBoundsException) {
            buffer.position(rollback)
            throw e
        }

        if (buffer.hasRemaining()) {
            listen(buffer)
        }
    }

    internal fun createStream(stream: RTMPStream) {
        call(
            "createStream",
            object : Responder {
                override fun onResult(arguments: List<Any?>) {
                    for (s in streams) {
                        if (s.value == stream) {
                            streams.remove(s.key)
                            break
                        }
                    }
                    val id = (arguments[0] as Double).toInt()
                    stream.id = id
                    streams[id] = stream
                    stream.readyState = RTMPStream.ReadyState.OPEN
                }
                override fun onStatus(arguments: List<Any?>) {
                    Log.w(javaClass.name + "#onStatus", arguments.toString())
                }
            }
        )
    }

    internal fun createConnectionMessage(): RTMPMessage {
        val paths = uri!!.path.split("/".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        val message = RTMPCommandMessage(RTMPObjectEncoding.AMF0)
        val commandObject = HashMap<String, Any?>()
        commandObject["app"] = paths[1]
        commandObject["flashVer"] = flashVer
        commandObject["swfUrl"] = swfUrl
        commandObject["tcUrl"] = uri!!.toString()
        commandObject["fpad"] = false
        commandObject["capabilities"] = RTMPConnection.DEFAULT_CAPABILITIES
        commandObject["audioCodecs"] = SupportSound.AAC.rawValue
        commandObject["videoCodecs"] = SupportVideo.H264.rawValue
        commandObject["videoFunction"] = VideoFunction.CLIENT_SEEK.rawValue
        commandObject["pageUrl"] = pageUrl
        commandObject["objectEncoding"] = objectEncoding.rawValue
        message.chunkStreamID = RTMPChunk.COMMAND
        message.streamID = 0
        message.commandName = "connect"
        message.transactionID = ++transactionID
        message.commandObject = commandObject
        if (arguments != null) {
            message.arguments = arguments
        }
        return message
    }

    companion object {
        const val DEFAULT_PORT = 1935
        const val DEFAULT_FLASH_VER = "FMLE/3.0 (compatible; FMSc/1.0)"
        val DEFAULT_OBJECT_ENCODING = RTMPObjectEncoding.AMF0

        private const val DEFAULT_CHUNK_SIZE_S = 1024 * 8
        private const val DEFAULT_CAPABILITIES = 239
    }
}
